src/conversation_engine/engine.py

```diff
@@ -308,102 +397,170 @@ def _handle_additional_info(self, text):
                                   "Anything to do boss?"])
-        example = self._parser.get_random_sentence(self._command_target)
 
-        sentence += " An example command is: '{}'. ".format(example)
+        if self.give_examples:
+            example = self._parser.get_random_sentence(self._command_target)
+
+            sentence += " An example command is: '{}'. ".format(example)
 
-        self._robot_to_user_pub.publish(sentence)
+        self._say_to_user(sentence)
 
     def _done_cb(self, task_outcome):
+        """The action_server's action is done, which can mean the action is finished (successfully or failed) or
+        needs additional info. This last option is handled by _on_request_missing_information and
+        the other cases by a starting a new conversation"""
         rospy.loginfo("_done_cb: Task done -> {to}".format(to=task_outcome))
         assert isinstance(task_outcome, TaskOutcome)
 
         self._latest_feedback = None
 
         if task_outcome.succeeded:
             rospy.loginfo("Action succeeded")
-            self._robot_to_user_pub.publish(" ".join(task_outcome.messages))
+            self._on_task_successful(" ".join(task_outcome.messages))
 
             self._start_new_conversation()
-
         elif task_outcome.result == TaskOutcome.RESULT_MISSING_INFORMATION:
             rospy.loginfo("Action needs more info from user")
 
             sentence = "".join(task_outcome.messages)
-            self._state.wait_for_user(target=self._get_grammar_target(task_outcome.missing_field),
+            target = self._get_grammar_target(task_outcome.missing_field)
+            self._state.wait_for_user(target=target,
                                       missing_field=task_outcome.missing_field)
-            example = self._parser.get_random_sentence(self._state.target)
-            sentence += " For example: '{}'".format(example)
-            self._robot_to_user_pub.publish(sentence)
+            self._on_request_missing_information(sentence, self._grammar, target)
 
         elif task_outcome.result == TaskOutcome.RESULT_TASK_EXECUTION_FAILED:
             rospy.loginfo("Action execution failed")
-            self._robot_to_user_pub.publish("".join(task_outcome.messages))
+
+            self._on_task_outcome_failed("".join(task_outcome.messages))
 
             self._start_new_conversation()
 
         elif task_outcome.result == TaskOutcome.RESULT_UNKNOWN:
             rospy.loginfo("Action result: unknown")
-            self._robot_to_user_pub.publish("".join(task_outcome.messages))
+            self._on_task_outcome_unknown("".join(task_outcome.messages))
 
             self._start_new_conversation()
 
         else:
             rospy.loginfo("Action result: other")
-            self._robot_to_user_pub.publish("".join(task_outcome.messages))
+            self._say_to_user("".join(task_outcome.messages))
 
             self._start_new_conversation()
```

Comment:
Strictly speaking, you can make this shorter since all but one case needs `_start_new_conversation()`

Reply:
@LoyVanBeek if this issue can be resolved, it is ok to merge.

Reply:
@jlunenburg not sure what you mean here?